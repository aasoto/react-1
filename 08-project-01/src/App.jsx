import { AuthProvider } from "./auth"
import { Navbar, Sidebar, SidebarProvider } from "./ui"

export const App = () => {
  return (
    <AuthProvider>
      <SidebarProvider>
        <Navbar />
        <Sidebar />
      </SidebarProvider>
      {/* <div className="bg-slate-800 w-full h-screen">
        <img src="images/miss-universe-logo.svg" alt="miss-universe-logo" />
        <h1 className="text-3xl text-white font-bold underline">
          Hello world!
        </h1>
      </div> */}
    </AuthProvider>
  )
}
