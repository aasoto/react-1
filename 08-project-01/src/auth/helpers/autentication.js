import { users } from "../data"

export const authentication = (email, password) => {
  return users.find(user => ((user.email === email) && (user.password === password)));
}