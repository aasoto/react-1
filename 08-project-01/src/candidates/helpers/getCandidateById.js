import { candidates } from "../data"

export const getCandidateById = (id) => {
  return candidates.find(candidate => candidate.id === id);
}