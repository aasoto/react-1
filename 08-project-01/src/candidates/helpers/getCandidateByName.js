import { candidates } from "../data"

export const getCandidateByName = (name) => {
  
  return candidates.filter( candidate => {
    return name && candidate.name.toLocaleLowerCase().includes(name);
  });

}