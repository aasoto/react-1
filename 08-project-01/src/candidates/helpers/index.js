export * from './getCandidateById';
export * from './getCandidateByName';
export * from './getCandidateByPlacement';
