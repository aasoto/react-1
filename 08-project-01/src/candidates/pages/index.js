export * from './HomePage';
export * from './router';