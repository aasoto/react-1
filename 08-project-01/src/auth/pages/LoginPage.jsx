import React, { useContext } from 'react'
import { useEffect } from 'react';
import { useState } from 'react';
import { useForm } from '../../hooks'
import { AlertErrorAuth } from '../components';
import { AuthContext } from '../context';
import { authentication } from '../helpers';

export const LoginPage = () => {

  const {login} = useContext(AuthContext);

  const [alertErrorAuth, setAlertErrorAuth] = useState(false);
  
  const {email, password, onInputChange} = useForm({
    email: '',
    password: ''
  });

  useEffect(() => {
    if (alertErrorAuth) {
      setAlertErrorAuth(() => !alertErrorAuth);
    }
  }, [email, password]);

  const onLoginSubmit = (event) => {
    event.preventDefault();
    if ((email.trim().length <= 1) || (password.length <= 6)) return;

    const autenticated = authentication(email, password);

    if(autenticated) {
      login(autenticated.name);

    } else {
      setAlertErrorAuth(() => !alertErrorAuth);
    }
  }

  return (
    <>
      <div className='flex justify-center items-center h-screen'>
        <div className='w-72 md:w-96 h-80 bg-white rounded-lg shadow-xl'>
          <div className='w-full h-16 rounded-t-lg text-center border-b border-gray-400 flex justify-center items-center'>
            <h1 className='text-rose-700 text-3xl font-bold'>Login</h1>
          </div>
          <AlertErrorAuth show={alertErrorAuth}/>
          <form onSubmit={onLoginSubmit}>
            <div className='flex flex-col justify-center items-center gap-5 px-8 pt-10 pb-4'>
              <input 
                className='w-full h-12 px-5 border rounded-md border-gray-400 placeholder:text-gray-400' 
                type="email" 
                placeholder='Email'
                name="email"
                aria-label='email'
                value={email}
                onChange={onInputChange}
              />
              <input 
                className='w-full h-12 px-5 border rounded-md border-gray-400 placeholder:text-gray-400' 
                type="password" 
                placeholder='Password'
                name="password"
                aria-label='input-password'
                value={password}
                onChange={onInputChange}
              />
            </div>
            <div className='flex justify-end items-center gap-3 p-8'>
              <button className='bg-rose-700 hover:bg-rose-800 text-white font-bold rounded-md px-10 py-2 scale-100 hover:scale-105 transition duration-200'>Ok</button>
            </div>
          </form>
        </div>
      </div>
    </>
  )
}
