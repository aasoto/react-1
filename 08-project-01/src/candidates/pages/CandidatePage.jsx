import { useState } from "react";
import { useContext } from "react";
import { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router-dom"
import { SidebarContext } from "../../ui";
import { getCandidateById } from "../helpers";

export const CandidatePage = () => {
  
  const {showSidebar, pathUp} = useContext(SidebarContext);
  
  pathUp();

  const {id} = useParams();

  const navigate = useNavigate();

  const candidate = useMemo(() => getCandidateById(id), [id]);

  const { name, age, country, instagram, meet, placement } = candidate;
  
  const onNavigateBack = () => {
    navigate(-1);
  }

  if ( !candidate ) {
    return <Navigate to="/" />
  }

  return (
    <>
      <div className={`pt-24 pr-5 animate__animated ${ showSidebar ? 'pl-80 animate__fadeInLeft' : 'pl-5 animate__fadeInRight' }`}>
        <h1 className="text-gray-600 font-semibold text-3xl md:text-4xl mb-2">{name}</h1>
        <hr className="border"/>
        <div className={`mt-5 grid grid-cols-1 ${ showSidebar ? 'md:grid-cols-3' : 'md:grid-cols-4'} gap-5`}>
          <div className="col-span-1">
            <img src={`./../images/candidates/${id}.PNG`} alt="" className="w-full object-cover object-top rounded-lg"/>
          </div>
          <div className={`${ showSidebar ? 'col-span-2' : 'col-span-3' }`}>
            <div className="grid grid-cols-3 gap-5">
              <div className="col-span-1">
                <span className={`fi fi-${id} scale-[6] md:scale-[10] translate-x-[53px] md:translate-x-[100px] translate-y-[40px] md:translate-y-[68px] rounded-sm`}></span>
              </div>
              <div className="col-span-2">
                <table className="w-full">
                  <tbody>
                    <tr>
                      <td className="text-gray-500 font-medium text-lg">
                        Country:
                      </td>
                      <td colSpan={2} className="text-gray-600 font-bold text-2xl">
                        {country.toUpperCase()}
                      </td>
                    </tr>
                    <tr>
                      <td className="text-gray-500 font-medium text-lg">
                        Age:
                      </td>
                      <td colSpan={2} className="text-gray-600 font-bold text-2xl">
                        {age}
                      </td>
                    </tr>
                    <tr>
                      <td className="text-gray-500 font-medium text-lg">
                        Instagram:
                      </td>
                      <td colSpan={2} className="text-gray-600 text-xl">
                        <a href={`https://instagram.com/${instagram.slice(1)}`}>
                          {instagram}
                        </a>
                      </td>
                    </tr>
                  </tbody>
                </table>  
              </div>
            </div>
            <h3 className="mt-20 text-gray-600 font-bold text-2xl">MEET</h3>
            <p className="text-gray-600 text-lg">{meet}</p>
            <button aria-label="btn-back" className="m-5 px-5 py-2 bg-rose-700 hover:bg-rose-800 text-white font-medium hover:font-semibold scale-100 hover:scale-105 shadow-none hover:shadow transition duration-200" onClick={onNavigateBack}>
              <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" strokeWidth={1.5} stroke="currentColor" className="w-6 h-6">
                <path strokeLinecap="round" strokeLinejoin="round" d="M9 15L3 9m0 0l6-6M3 9h12a6 6 0 010 12h-3" />
              </svg>
              Back
            </button>
          </div>
        </div>
      </div>
    </>
  )
}
