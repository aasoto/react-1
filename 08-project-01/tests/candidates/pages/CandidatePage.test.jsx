import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { AuthContext } from "../../../src/auth";
import { getCandidateById } from "../../../src/candidates/helpers/getCandidateById";
import { CandidatePage } from "../../../src/candidates/pages/CandidatePage";
import { SidebarContext } from "../../../src/ui";

/** AQUI SE ESTÁ UTILIZANDO UN HELPER */
//se debe hacer un mock de todo el archivo del helper
jest.mock('../../../src/candidates/helpers/getCandidateById');

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
})); 

const showSidebarMock = jest.fn();
const pathUpMock = jest.fn();


describe('Pruebas de <CandidatePage/>', () => {

  /** se debe dar el valor que debería retornar con la url dada */
  getCandidateById.mockReturnValue({
    'id': 'co',
    'name': 'MARÍA FERNANDA ARISTIZÁBAL',
    'country': 'Colombia',
    'age': '25',
    'instagram': '@mafearistizabalu',
    'meet': 'María Fernanda Aristizabál is a motivational speaker, model, and influencer with a degree in media and journalism. Through her campaign #MakeEveryMomentWorthIt, she shares her philosophy of life via supporting vulnerable communities in Colombia. She is especially passionate about generating a positive impact in children’s mental health. Aristizabal works with various nonprofit organizations, including Smile Train and Best Buddies. Aristizabal wishes to use her Miss Universe platform to give visibility and awareness to water scarcity through her campaign #SOSWATER.',
    'placement': 'semi-finalist'
  });

  beforeEach(() => jest.clearAllMocks());

  test('debe mostrarse correctamente el snapshot', () => {

    const {container} = render(
      <AuthContext.Provider value={{logged: true}}>
        <MemoryRouter>
          <SidebarContext.Provider value={{showSidebar: showSidebarMock, pathUp: pathUpMock}}>
            <CandidatePage/>
          </SidebarContext.Provider>
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect( container ).toMatchSnapshot();

  });

  test('debe llamar a la función onNavigateBack()', () => {

    render(
      <AuthContext.Provider value={{logged: true}}>
        <MemoryRouter initialEntries={['/candidate/co']}>
          <SidebarContext.Provider value={{showSidebar: showSidebarMock, pathUp: pathUpMock}}>
            <CandidatePage/>
          </SidebarContext.Provider>
        </MemoryRouter>
      </AuthContext.Provider>
    );

    const btnBack = screen.getByLabelText('btn-back');
    fireEvent.click(btnBack);

    expect(mockedUseNavigate).toHaveBeenCalledWith(-1);
  });

});