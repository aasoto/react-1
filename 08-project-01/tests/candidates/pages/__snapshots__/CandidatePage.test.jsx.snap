// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <CandidatePage/> debe mostrarse correctamente el snapshot 1`] = `
<div>
  <div
    class="pt-24 pr-5 animate__animated pl-80 animate__fadeInLeft"
  >
    <h1
      class="text-gray-600 font-semibold text-3xl md:text-4xl mb-2"
    >
      MARÍA FERNANDA ARISTIZÁBAL
    </h1>
    <hr
      class="border"
    />
    <div
      class="mt-5 grid grid-cols-1 md:grid-cols-3 gap-5"
    >
      <div
        class="col-span-1"
      >
        <img
          alt=""
          class="w-full object-cover object-top rounded-lg"
          src="./../images/candidates/undefined.PNG"
        />
      </div>
      <div
        class="col-span-2"
      >
        <div
          class="grid grid-cols-3 gap-5"
        >
          <div
            class="col-span-1"
          >
            <span
              class="fi fi-undefined scale-[6] md:scale-[10] translate-x-[53px] md:translate-x-[100px] translate-y-[40px] md:translate-y-[68px] rounded-sm"
            />
          </div>
          <div
            class="col-span-2"
          >
            <table
              class="w-full"
            >
              <tbody>
                <tr>
                  <td
                    class="text-gray-500 font-medium text-lg"
                  >
                    Country:
                  </td>
                  <td
                    class="text-gray-600 font-bold text-2xl"
                    colspan="2"
                  >
                    COLOMBIA
                  </td>
                </tr>
                <tr>
                  <td
                    class="text-gray-500 font-medium text-lg"
                  >
                    Age:
                  </td>
                  <td
                    class="text-gray-600 font-bold text-2xl"
                    colspan="2"
                  >
                    25
                  </td>
                </tr>
                <tr>
                  <td
                    class="text-gray-500 font-medium text-lg"
                  >
                    Instagram:
                  </td>
                  <td
                    class="text-gray-600 text-xl"
                    colspan="2"
                  >
                    <a
                      href="https://instagram.com/mafearistizabalu"
                    >
                      @mafearistizabalu
                    </a>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <h3
          class="mt-20 text-gray-600 font-bold text-2xl"
        >
          MEET
        </h3>
        <p
          class="text-gray-600 text-lg"
        >
          María Fernanda Aristizabál is a motivational speaker, model, and influencer with a degree in media and journalism. Through her campaign #MakeEveryMomentWorthIt, she shares her philosophy of life via supporting vulnerable communities in Colombia. She is especially passionate about generating a positive impact in children’s mental health. Aristizabal works with various nonprofit organizations, including Smile Train and Best Buddies. Aristizabal wishes to use her Miss Universe platform to give visibility and awareness to water scarcity through her campaign #SOSWATER.
        </p>
        <button
          aria-label="btn-back"
          class="m-5 px-5 py-2 bg-rose-700 hover:bg-rose-800 text-white font-medium hover:font-semibold scale-100 hover:scale-105 shadow-none hover:shadow transition duration-200"
        >
          <svg
            class="w-6 h-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 15L3 9m0 0l6-6M3 9h12a6 6 0 010 12h-3"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          Back
        </button>
      </div>
    </div>
  </div>
</div>
`;
